src/refract/Element.h

```diff
@@ -2,798 +2,186 @@
-        {
-        }
-
-        RefElement(const TraitType::ValueType& value) : Type()
-        {
-            set(value);
-        }
-    };
-
-    struct ExtendElementTrait : public ElementCollectionTrait<> {
-        static const std::string element()
-        {
-            return "extend";
-        }
-    };
-
-    struct ExtendElement : Element<ExtendElement, ExtendElementTrait> {
-        ExtendElement() : Type()
-        {
-        }
-
-        ExtendElement(const TraitType::ValueType& value) : Type()
-        {
-            set(value);
-        }
-
-        void push_back(IElement* e);
-        void set(const ValueType& val);
-
-        IElement* merge() const;
-    };
-
-    struct OptionElementTrait : public ElementCollectionTrait<> {
-        static const std::string element()
-        {
-            return "option";
-        }
-    };
-
-    struct OptionElement : Element<OptionElement, OptionElementTrait> {
-
-        OptionElement() : Type()
-        {
-        }
-
-        OptionElement(const TraitType::ValueType& value) : Type()
-        {
-            set(value);
-        }
-
-        OptionElement(IElement* e) : Type()
-        {
-            value.push_back(e);
-        }
-
-        void push_back(IElement* e)
-        {
-            // FIXME:
-            // probably add check for allowed type
-            hasContent = true;
-            value.push_back(e);
-        }
-    };
-
-    struct SelectElementTrait : public ElementCollectionTrait<OptionElement> {
-        static const std::string element()
-        {
-            return "select";
-        }
-
-        static void cloneValue(const ValueType& self, ValueType& other)
-        {
-            for (std::vector<OptionElement*>::const_iterator it = self.begin(); it != self.end(); ++it) {
-                IElement* clonned = (*it)->clone();
-                other.push_back(static_cast<OptionElement*>(clonned));
-            }
-        }
-    };
-
-    struct SelectElement : Element<SelectElement, SelectElementTrait> {
-        SelectElement() : Type()
-        {
-        }
+    ///
+    /// Create an empty Element of given type
+    /// @remark an empty Element has an empty data structure definition (DSD)
+    ///
+    template <typename ElementT>
+    auto make_empty()
+    {
+        return std::make_unique<ElementT>();
+    }
+
+    ///
+    /// Create an Element of given type forwarding arguments to the constructor
+    /// of its data structure definition (DSD)
+    ///
+    template <typename ElementT, typename... Args>
+    auto make_element(Args&&... args)
+    {
+        using DataT = typename ElementT::ValueType;
+        return std::make_unique<ElementT>(DataT{ std::forward<Args>(args)... });
+    }
 
-        SelectElement(const TraitType::ValueType& value) : Type()
-        {
-            set(value);
-        }
+    template <typename Primitive, typename DataT = typename data::data_of<Primitive>::type>
+    auto make_primitive(const Primitive& p)
```

Comment:
And another one proposal. to avoid calling like this
```
make_primitive(static_cast<double>(annotation.code));
```
Add into DSD definition of supported "conversion" like:
* number -> std::is_numeric
* string -> std::string, const char*

Reply:
I can implement number from `int`, but I think we should use some big number representation that implicitly constructs from `int`, `double`, etc.

Reply:
Probably best solution is hold it as string internal.
There are other problem like parsing of localized numbers, thousand separators, decimal point, transformation to different format etc.

Reply:
On second thought, I don't think it helps to bury the static_cast in a helper. I think the user of `make_primitive` should take care of providing a `double` as long as we don't provide a bignum constructor.

Reply:
Imagine `make_primitive` accepts an `int`. The user of that function has no clue there are restrictions on the `int` because it is cast into a `double` somewhere in the guts.